api/src/transactions.rs

```diff
@@ -209,6 +210,73 @@ impl TransactionsApi {
             .await
     }
 
+    /// Wait for transaction by hash
+    ///
+    /// Same as /transactions/by_hash, but will wait for a pending transaction to be committed. To be used as a long
+    /// poll optimization by clients, to reduce latency caused by polling. The "long" poll is generally a second or
+    /// less but dictated by the server; the client must deal with the result as if the request was a normal
+    /// /transactions/by_hash request, e.g., by retrying if the transaction is pending.
+    #[oai(
+        path = "/transactions/wait_by_hash/:txn_hash",
+        method = "get",
+        operation_id = "wait_transaction_by_hash",
+        tag = "ApiTags::Transactions"
+    )]
+    async fn wait_transaction_by_hash(
+        &self,
+        accept_type: AcceptType,
+        /// Hash of transaction to retrieve
+        txn_hash: Path<HashValue>,
+        // TODO: Use a new request type that can't return 507.
+    ) -> BasicResultWith404<Transaction> {
+        fail_point_poem("endpoint_wait_transaction_by_hash")?;
+        self.context
+            .check_api_output_enabled("Get transactions by hash", &accept_type)?;
+
+        // Short poll if the active connections are too high
+        if self
+            .context
+            .wait_for_hash_active_connections
+            .load(std::sync::atomic::Ordering::Relaxed)
+            >= self
+                .context
+                .node_config
+                .api
+                .wait_by_hash_max_active_connections
+        {
+            metrics::WAIT_TRANSACTION_POLL_TIME
+                .with_label_values(&["short"])
+                .observe(0.0);
+            return self
+                .get_transaction_by_hash_inner(&accept_type, txn_hash.0)
+                .await;
+        }
+
+        let start_time = std::time::Instant::now();
+        self.context
+            .wait_for_hash_active_connections
+            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
```

Comment:
Hmm there is a race condition here as we increment after checking the counter. I think it is okay as we don't have to guarantee that the active connections is strictly less than the threshold but let's add a comment here stating that is intentional.